### PR TITLE
[SDPA] Create stub function for doing SDPA cpp and cuda dispatch

### PR DESCRIPTION
**Pull request by drisspg**
## Summary
Torch.compile was previously not working for transformerencoder because torch.SDPA calls a native function on tensors that returns an int. This PR instead creates a dispatch stub for the function called in order to not create a separate fx node for this native function.
As well this pr adds meta functions for the fused kerenels.

### CODE REVIEW

**Review by cpuhrsch — APPROVED** [review]

**Comment by drisspg** [comment]
torch.compile(model, full_graph=True)
Is successfully able to compile with no graph breaks. It did create a fallback for sdpa which is expected and currently desired.

```
class FlashMHA(torch.nn.Module):
    def __init__(self, num_heads, qkv, proj):
        super().__init__()
        self.qkv = qkv
        self.proj = proj
        self.num_heads = num_heads

    def forward(self, query: torch.Tensor, key: torch.Tensor, value: torch.Tensor):
        # assert query is key
        # assert key is value
        # Asserts don't play well. As well as sdp_kernel() is not playing well

        query_projected = torch.nn.functional.linear(
            query, self.qkv.weight, self.qkv.bias)

        batch_size = query_projected.size(0)
        seq_len = query_projected.size(1)
        embed_dim = query_projected.size(2)
        head_dim = embed_dim//(self.num_heads*3)

        # Transpose seq_len and num_heads dim
        query_projected = query_projected.view(batch_size, seq_len, 3*self.num_heads, head_dim).transpose(1, 2)
        query, key, value = query_projected.chunk(3, 1)

        # the output of sdp = (batch, num_heads, seq_len, head_dim)
        attn, _ = torch.nn.functional._scaled_dot_product_attention(
            query, key, value, attn_mask=None, dropout_p=0.0, need_attn_weights=False, is_causal=False)

        # Determine which one of these is faster
        attn = attn.transpose(1, 2).view(batch_size, seq_len, self.num_heads*head_dim)
        # attn = attn.transpose(1, 2).contiguous().view(batch_size, seq_len, self.num_heads*head_dim)
        attn = torch.nn.functional.linear(
            attn, self.proj.weight, self.proj.bias)
        return attn, None
     ```
     
  I have another script that succeeds on the forward but inductor breaks on the backward compilation. This is a full TransformerEncoder.

**Comment by drisspg** [comment]
The above not withstanding I think that this PR is okay to land and is a meaningful step in making sdp more PT2.0 friendly. 
Follow ups:
1.  Does torch re-compile model when the at::Context changes causing the a different sdp function to be returned for the same inputs.
2. Figure out why AOT autograd is not producing a correct backward graph and causing errors. ( I suspect this has to do with how SDP_efficient optimizes for chunked QKV).

**Comment by drisspg** [comment]
Some more errors. The dispatch stub does not support meta tensors or xla. Might try to just add some inline dispatching on query.device()

**Comment by eellison** [comment]
So, if all the properties needed to calculate `select_sdp_backend` are known statically, we could make this a constant and add the appropriate guards on the inputs. However, since we're not going to fuse the attention kernel, it might be easeir to just graph break. 

For static shapes, adding the guards is easy enough, but for dynamic shapes it'll be more complicated since the kernel we run is [dependent on the input sizes](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/transformers/cuda/sdp_utils.h#L147).  I think we'd need to do something similar to @ezyang's [conv pr](https://github.com/pytorch/pytorch/pull/89069).

cc @yanboliang, @ngimel

**Comment by eellison** [comment]
Fake Tensor has basically two modes, one in which it reports its fake device, and one in which it reports its meta device. When it runs the meta kernel, it defaults to using the meta device because otherwise the meta kernels would error in weird ways. 

So, what you're seeing is it's running a kernel and reporting meta. We had a similar problem with Conv (which has very complex logic to decide what backend to run, and those backends have different striding behavior).

Here is the fix for Conv : https://github.com/pytorch/pytorch/pull/87305/. We ran the logic for deciding the backend with FakeTensor in the mode that reports CUDA/CPU., and then have a separate function that given the backend and arguments will determine the output memory striding.

I would recommend doing something similar here.

File: aten/src/ATen/native/transformers/attention.h [hunk 1]
```diff
@@ -1,10 +1,17 @@
 #pragma once
 #include <ATen/ATen.h>
 #include <c10/macros/Export.h>
+#include <ATen/native/DispatchStub.h>
+#include <ATen/native/transformers/attention.h>
```
Comment (lms-mt) [inline aten/src/ATen/native/transformers/attention.h:5]: Is this file self-included?

**Pull request closed without merging**